### PR TITLE
Adapting ct ng scripts

### PR DESCRIPTION
**Pull request by anthony-cros**
will comment on the code a bit!

### CODE REVIEW

File: Dockerfile-xcompiler [hunk 1]
```diff
@@ -33,6 +27,8 @@ RUN apt-get install -y \
    wget=1.15-1ubuntu1.14.04.1
 
 # External packages required by crosstool-ng
+RUN apt-get update
+RUN apt-get install --fix-missing
```
Comment (anthony-cros) [inline Dockerfile-xcompiler]: the problem i mentioned earlier

File: Dockerfile-xcompiler [hunk 2]
```diff
@@ -55,4 +51,3 @@ ADD cross-build/ /home/xcompiler/webthree-umbrella/cross-build/
 
 # Build the cross-compiler
 WORKDIR /home/xcompiler/webthree-umbrella/cross-build/ct-ng
-RUN ./ct-ng.sh ~/ct-ng $TARGET_ARCHITECTURE "1.20.0"
```
Comment (anthony-cros) [inline Dockerfile-xcompiler]: newer versions of docker seem to have an [ARG](http://docs.docker.com/engine/reference/builder/#arg) command. unless we update to it, we won't be able to parameterize the docker file it seems, so I went with [this suggestion](https://github.com/docker/docker/issues/14634#issuecomment-136793152) instead, it's pretty clean IMO

File: build-xcompiler-armel.sh [hunk 3]
```diff
@@ -1 +1,12 @@
-TARGET_ARCHITECTURE="arm-unknown-linux-gnueabi" docker build -f Dockerfile-xcompiler -t xcompiler .
+#!/bin/bash
+docker build \
```
Comment (anthony-cros) [inline build-xcompiler-armel.sh]: builld "base" image

File: build-xcompiler-armel.sh [hunk 4]
```diff
@@ -1 +1,12 @@
...
+  -t xcompiler \
+  -f Dockerfile-xcompiler \
+  .
+docker run \
```
Comment (anthony-cros) [inline build-xcompiler-armel.sh]: build the customized one, note that we could commit it afterward

File: cross-build/ct-ng/ct-ng.sh [hunk 5]
```diff
@@ -4,15 +4,19 @@
...
+source ../ethereum/utils.sh
 
 # TODO: enforce both or none
-CONFIG_CHANGES=${1?} && shift # must provide a config file path or "default"; 
 TARGET_ARCHITECTURE=$1
 TARGET_ARCHITECTURE=${TARGET_ARCHITECTURE:="arm-unknown-linux-gnueabi"}
 CTNG_VERSION=$2
 CTNG_VERSION=${CTNG_VERSION:="1.20.0"}
 
+readonly DEFAULT_CONFIG_CHANGES_VALUE="none"
```
Comment (anthony-cros) [inline cross-build/ct-ng/ct-ng.sh]: changed from "default" to "none" (since I'm referring to config changes)

File: cross-build/ct-ng/ct-ng.sh [hunk 6]
```diff
@@ -26,52 +30,8 @@ readonly CROSS_COMPILER_DIR="$HOME/x-tools/${TARGET_ARCHITECTURE?}" # TODO: how
 readonly DEPENDENCIES="" #bison flex texinfo gawk automake libtool cvs ncurses-dev gperf" # TODO: re-enable once dependencies are fleshed out
 hash ${DEPENDENCIES?} 2>&- || { echo -e "ERROR: missing one or more dependencies amongst:\n\t${DEPENDENCIES?}"; exit 1; }
 /sbin/ldconfig -p | grep libexpat || { echo "ERROR: libexpat is missing, please run 'sudo apt-get install libexpat1-dev'"; exit 1; }
-[ "${CONFIG_CHANGES?}" == "default" ] || [ -f "${CONFIG_CHANGES?}" ] ||  { echo "ERROR: must provide config file changes or the value \"default\"" exit 1; }
-
-# ===========================================================================
-
-# TEMPORARILY BORROWED FROM ethereum dir's utils.sh, DO NOT EDIT generic_hack
```
Comment (anthony-cros) [inline cross-build/ct-ng/ct-ng.sh]: removed duplicate

File: cross-build/ct-ng/ct-ng.sh [hunk 7]
```diff
@@ -26,52 +30,8 @@ readonly CROSS_COMPILER_DIR="$HOME/x-tools/${TARGET_ARCHITECTURE?}" # TODO: how
...
-  
-  if [ ! -f ${FILE?}.bak ]; then
-    cp ${FILE?} ${FILE?}.bak # to keep original file around conveniently (without having to use git)
-  fi
-  
-  awk "${AWK_SCRIPT?}" ${FILE?}.bak > ${FILE?}
-}
-
-# ---------------------------------------------------------------------------
-function to_awk_script() {
```
Comment (anthony-cros) [inline cross-build/ct-ng/ct-ng.sh]: moved to utils to unclutter the script

**Comment by anthony-cros** [comment]
done commenting, your turn!

**Comment by anthony-cros** [comment]
did you make changes since this morning? looks like we have a conflict

**Comment by bobsummerwill** [comment]
Great - please proceed and commit!    Just pull from master first to resolve the clash.   It's likely you'll want to stomp my edits anyway.

**Comment by bobsummerwill** [comment]
The Docker ARGS thing is weird, because the exact same pattern worked for Dockerfile-eth.

**Comment by bobsummerwill** [comment]
So yeah - don't wait for me, @anthony-cros :-)
Thanks for the PR and detail, but you don't need my OK to commit!

**Comment by bobsummerwill** [comment]
This is a good read:

    http://crosbymichael.com/dockerfile-best-practices.html

See especially the bit at the end about ENTRYPOINT.    This really matches how we are using Docker.

**Comment by bobsummerwill** [comment]
The other Docker TODO which you could roll into this is "fixing" the "copying the files out" step.
Perhaps the best way of doing that would be flipping those final commands (xcompiler build or eth build) to be ENTRYPOINT, so that both process are ...

docker build (does everything EXCEPT the actual build process)
docker run (does the build process and copies files out).

karalabe said the way to do that was by mounting a volume from the host into the container, and then copying to that.

**Comment by anthony-cros** [comment]
> The Docker ARGS thing is weird, because the exact same pattern worked for Dockerfile-eth.

what do you mean? Our Dockerfile-eth doesn't seem to use ARG

**Comment by bobsummerwill** [comment]
Not ARG, no, but ENV.

**Comment by anthony-cros** [comment]
I think ENV works if it is set in the Dockerfile itself, but we can't just export a EV outside of it and have it be "picked up"

**Comment by bobsummerwill** [comment]
Right - so what I thought was defaulting wouldn't actually have worked for overrides.   Gotcha.

**Pull request merged**